src/server/Controllers/StoresController.cs

```diff
@@ -149,59 +144,42 @@ public async Task<IActionResult> GetStoreCatchAll(string storeId, string catchAl
                 return Redirect($"{referer}/{storeId}/{key}");
             }
 
+            // Requesting GetValue only from the last root hash onchain ensures that only
+            // nodes that have the latest state will respond to the request
+            // This helps prevent a mismatch between the state of the store and
+            // the data when pulled across decentralized nodes
+            var lastRootHash = await _gatewayService.GetLastRoot(storeId, cancellationToken);
+            if (lastRootHash is null)
+            {
+                return NotFound();
+            }
+
             // info.html is a synthetic key that we use to display the store's contents
             // even though index.html returns the same, if the store overrides index.html
             // the user can still get a list of keys at info.html
             if (key == "info.html")
             {
-                var keys = await _gatewayService.GetKeys(storeId, cancellationToken);
+                var keys = await _gatewayService.GetKeys(storeId, lastRootHash, cancellationToken);
                 if (keys is not null)
                 {
-                    var decodedKeys = keys.Select(HexUtils.FromHex).ToList();
-                    string htmlContent = IndexRenderer.Render(storeId, decodedKeys, Request);
+                    var htmlContent = IndexRenderer.Render(storeId, keys.Select(HexUtils.FromHex));
 
                     return Content(htmlContent, "text/html");
                 }
             }
 
             var hexKey = key.StartsWith("0x") ? key : HexUtils.ToHex(key);
```

Comment:
should probably put a comment here just saying the point of this check is to support requesting keys by hex or utf8